Introducing ClientRequestFilter.java: A New Plugin for Merging Additional Headers into Client Requests in the Authentication Filter (#23380)
prestodb/presto: presto-docs/src/main/sphinx/develop/client-request-filter.rst

======================
Client Request Filter
======================

> Presto allows operators to customize the headers used by the Presto runtime to process queries. Some example use cases include customized authentication workflows, or enriching query attributes such as the query source. This can be achieved using the Client Request Filter plugin, which allows control over header customization during query execution.

Reviewer: ```suggestion
Presto allows operators to customize the headers used to process queries. Some example use cases include customized authentication workflows, or enriching query attributes such as the query source. Use the Client Request Filter plugin to control header customization during query execution.
```
Some suggestions for simpler phrasing and conciseness.

Implementation
--------------

The ``ClientRequestFilterFactory`` is responsible for creating instances of ``ClientRequestFilter``. It also defines 
the name of the filter.

The ``ClientRequestFilter`` interface provides two methods: ``getExtraHeaders()``, which allows the runtime to quickly check if it needs to apply a more expensive call to enrich the headers, and ``getHeaderNames()``, which returns a list of header names used as the header names in client requests.

The implementation of ``ClientRequestFilterFactory`` must be wrapped as a plugin and installed on the Presto cluster.

After installing a plugin that implements ``ClientRequestFilterFactory`` on the coordinator, the ``AuthenticationFilter`` class passes the ``principal`` object to the request filter, which returns the header values as a map.

> Presto uses the request filter to determine whether a header is present in the blocklist. The blocklist includes headers such as ``X-Presto-Transaction-Id``, ``X-Presto-Started-Transaction-Id``, ``X-Presto-Clear-Transaction-Id``, and ``X-Presto-Trace-Token``, which are not allowed to be overridden. For a complete list of headers that may be overridden, please refer to the `Java source`_ (note that the blocklist headers are also listed there, but they are not eligible for overriding).

Reviewer: ```suggestion
Presto uses the request filter to determine whether a header is present in the blocklist. The blocklist includes headers such as ``X-Presto-Transaction-Id``, ``X-Presto-Started-Transaction-Id``, ``X-Presto-Clear-Transaction-Id``, and ``X-Presto-Trace-Token``, which are not allowed to be overridden. 

For a complete list of headers, see the `Java source`_. 

Note: The `Java source`_ includes these blocklist headers that are not eligible for overriding. The other headers not mentioned here can be overridden.
```
Some rephrasing, and adding line breaks for ease of reading. 

I made the assumption in my draft suggestion here that the headers you list are the complete list of the headers that cannot be overridden, and that the other headers in the Java source can all be overridden. If my assumption is wrong, please revise as you feel correct.

======================
Client Request Filter
======================

Presto facilitates seamless integration of additional headers into client requests using the 

> Client Request Filter plugin via the AuthenticationFilter class, enabling efficient and dynamic header management.

Author: Sure.

Implementation
--------------

The ``ClientRequestFilterFactory`` is responsible for creating instances of ``ClientRequestFilter``. It also defines 

> the name of the filter, which is subsequently used as the header name in client requests.

Author: Corrected.

``ClientRequestFilter``  interface provides two methods: ``getExtraHeaders()``,

> which returns header values as key-value pairs in a map, and ``getHeaderNames()``, which returns a list of header names used as the header names in client requests.

Reviewer 2: I imagine `getExtraHeaders` is used for the runtime to quickly check if it needs to apply a more expensive call to enrich the headers.  If so, add this context to the documentation.

Author: Added.

which returns header values as key-value pairs in a map and ``getHeaderNames()`` which returns a list of header names.

The implementation of ``ClientRequestFilterFactory`` must be wrapped
as a plugin and installed on the Presto cluster.

After installing a plugin that implements ``ClientRequestFilterFactory`` on the coordinator, the ``AuthenticationFilter`` class passes the ``principal`` object to the request filter, which returns the header values as a map.

> Presto uses the name of the request filter to determine whether the header is present in the blocklist.

Reviewer 2: What blocklist?  Mention the headers which are not allowed to be overidden.  Link to the Java source of headers that may be overidden.

Author: Added.
